### PR TITLE
Better handling of task collection failure

### LINKED ISSUE #1234: Make task collection cleanup more robust upon collection failure (related to "Pinned packages versions")
I've run into some edge-cases recently during task collection. For example: I mistyped the torch version, resulting in this collection error:

```
2024-02-20 10:10:29,554 - fractal-tasks-core - DEBUG - 
2024-02-20 10:10:29,554 - fractal-tasks-core - DEBUG - Currently installed version of torch (2.0.0) differs from pinned version (0.12.0); install version 0.12.0.
2024-02-20 10:10:30,691 - fractal-tasks-core - DEBUG - Subprocess call to: /path/to/FRACTAL_TASK_DIR/.fractal/fractal-tasks-core0.14.2/venv/bin/pip install torch==0.12.0
2024-02-20 10:10:30,691 - fractal-tasks-core - DEBUG - 
2024-02-20 10:10:30,691 - fractal-tasks-core - DEBUG - ERROR: Could not find a version that satisfies the requirement torch==0.12.0 (from versions: 1.7.1, 1.8.0, 1.8.1, 1.9.0, 1.9.1, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 2.0.0, 2.0.1, 2.1.0, 2.1.1, 2.1.2, 2.2.0)
ERROR: No matching distribution found for torch==0.12.0

2024-02-20 10:10:30,691 - fractal-tasks-core - ERROR - Task manifest loading failed
2024-02-20 10:10:30,691 - fractal-tasks-core - DEBUG - Task-collection status: fail
2024-02-20 10:10:30,691 - fractal-tasks-core - INFO - Background collection failed. Original error: ERROR: Could not find a version that satisfies the requirement torch==0.12.0 (from versions: 1.7.1, 1.8.0, 1.8.1, 1.9.0, 1.9.1, 1.10.0, 1.10.1, 1.10.2, 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 2.0.0, 2.0.1, 2.1.0, 2.1.1, 2.1.2, 2.2.0)
ERROR: No matching distribution found for torch==0.12.0
```

That seems fine, so let me try to actually install with torch 1.12.0. Unfortunately, the cleanup upon hitting such an error doesn't seem to run fully, resulting in the following error when trying to collect the package again with the correct torch dependency:

```
There has been an error, reason:
{
  "detail": "Cannot collect package. Possible reason: another collection of the same package is in progress. Original error: [Errno 2] No such file or directory: '/Path/to/FRACTAL_TASK_DIR/.fractal/fractal-tasks-core0.14.2/collection.json'"
}
```

Manually removing the `fractal-tasks-core0.14.2` folder then solves things.

### PR DESCRIPTION
**Pull request by ychiucco**
closes #1234 

## Checklist before merging
- [x] I added an appropriate entry to `CHANGELOG.md`
- [x] I merged `main` into the current branch.

### CODE REVIEW

File: tests/v2/03_api/test_api_task_collection_failures.py [hunk 1]
```diff
@@ -178,3 +175,56 @@ async def test_collection_validation_error(
...
+    async with MockCurrentUser(user_kwargs=dict(is_verified=True)):
+
+        res = await client.post(
+            f"{PREFIX}/collect/pip/",
+            json=dict(
+                **payload, pinned_package_versions={"devtools": "99.99.99"}
+            ),
+        )
+        assert res.status_code == 201
+        assert os.path.isdir(DIRECTORY) is False
```
Comment (tcompa) [inline tests/v2/03_api/test_api_task_collection_failures.py:214]: Comment everything starting from here

File: fractal_server/tasks/v2/background_operations.py [hunk 2]
```diff
@@ -358,14 +357,12 @@ async def background_collect_pip(
             # Write last logs to file
             logger.debug("Task-collection status: OK")
             logger.info("Background task collection completed successfully")
-            close_logger(logger)
```
Comment (tcompa) [inline fractal_server/tasks/v2/background_operations.py:361]: revert

File: fractal_server/tasks/v2/background_operations.py [hunk 3]
```diff
@@ -358,14 +357,12 @@ async def background_collect_pip(
...
             db.close()
 
         except Exception as e:
             # Write last logs to file
             logger.debug("Task-collection status: fail")
             logger.info(f"Background collection failed. Original error: {e}")
-            close_logger(logger)
```
Comment (tcompa) [inline fractal_server/tasks/v2/background_operations.py:368]: revert

**Comment by ychiucco** [comment]
`test_remove_directory` fails, but succeeds when executed alone (without the other tests in `tests/v2/03_api/test_api_task_collection_failures.py`).
In particular, the other test needed for it to fail is `test_missing_task_executable`:
```sh
$ poetry run pytest \
tests/v2/03_api/test_api_task_collection_failures.py::test_missing_task_executable \
tests/v2/03_api/test_api_task_collection_failures.py::test_remove_directory

...
...
...
======================================================================== short test summary info ========================================================================
FAILED tests/v2/03_api/test_api_task_collection_failures.py::test_remove_directory - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/j7/4knlv_bj7w37nc6vh_v76jgw0000gn/T/pytest-of-yuri/pytest-18/test_missing_task_executa...
=============================================================== 1 failed, 1 passed, 3 warnings in 20.59s ================================================================
```

**Pull request merged**